Controller/AbstractCrudController.php

```diff
@@ -165,6 +168,25 @@ private function getEntityValidator()
      */
     private function getDataMapper()
     {
+        $config = $this->getPreparedConfig();
+
+        /**
+         * !!! Please note that service Id of $config['data_mapper'] will be different from
+         * $this->container->get('data_mapper'];.
+         *
+         * $config['data_mapper'] is set inside Controller getConfig, and do not affect/overwrite service that defined
+         * in modera_server_crud.data_mapping.default_data_mapper service
+         *
+         * That allow to use custom data mapper and default data mapper simultaneously.
+         */
+        if (array_key_exists('data_mapper', $config) && $config['data_mapper']) {
+            if (!$this->container->has($config['data_mapper'])) {
+                throw new \RuntimeException("'data_mapper' configuration property exists, but no service with such id found.");
+            }
+
+            return $this->container->get($config['data_mapper']);
+        }
+
```

Comment:
The whole block can be simplified and refactored to something like this:

```
/**
 * @return \Modera\ServerCrudBundle\DataMapping\DataMapperInterface
 */
private function getDataMapper()
{
    $config = $this->getPreparedConfig();

    return call_user_func($config['create_default_data_mapper'], $this->container);
}
```

Then **getPreparedConfig** method can be updated to add this entry to $defaultConfig:

```
// allows to override default data mapper used by the this specific controller
'create_default_data_mapper' => function(ContainerInterface $container) {
    return $this->getConfiguredService('data_mapper');
}
```

With this implementation if it is needed developer can create a custom data-mapper even without registering in a service container.